lms/djangoapps/courseware/management/commands/migrate_student_module_history.py

```diff
@@ -20,63 +28,125 @@ class Command(BaseCommand):
+            help='delete the lock table, then exit. Run this after the migration is complete.')
+        make_option('-m' '--min-id', type='int', dest='min_id',
+            help='first (smallest) ID to migrate')
+        make_option('-M', '--max-id', type='int', dest='max_id',
+            help='last (greatest) id to migrate')
     )
 
     def handle(self, *arguments, **options):
-        if options['index'] >= options['num_chunks']:
-            self.stdout.write("Index {} is too large for {} chunks\n".format(options['index'], options['num_chunks']))
-            exit(2)
+        try:
+            StudentModuleHistoryArchive.objects.all()[0]
+        except IndexError:
+            self.stdout.write("No entries found in StudentModuleHistoryArchive, aborting migration.\n")
+            exit(1)
+
+        self.min_id = options['min_id']
+        self.max_id = options['max_id']
 
-        #Set max and min id from the number of chunks and the selected index
-        chunk_size = SQL_MAX_INT / options['num_chunks']
-        min_id = chunk_size * options['index']
+        initialized = self._check_initialized()
 
-        if options['index'] == options['num_chunks'] - 1:
-            max_id = SQL_MAX_INT
+        if options['cleanup']:
+            self._cleanup():
+            return
+        elif options['initialize']:
+            if initialized:
+                self.stdout.write("Migration is already initialized\n")
+                return
+            else:
+                self._initialize()
+                return
         else:
-            max_id = chunk_size * options['index'] + chunk_size - 1
+            if initialized:
+                self._migrate()
+                return
+            else:
+                self.stderr.write('The migration is not initialized. Run this command with "--initialize", then try again\n')
+                exit(2)
+
+
+    @transaction.commit_manually    #So that when we log a success message, there's no chance that it actually failed
+    def _initialize():
+        '''Initialize the lock table for the migration'''
+        logging.info("Initializing migration (creating lock table)...")
+        with closing(connection.cursor()):
+            try:
+                cursor.execute(dedent('''
+                    CREATE TABLE %s (
+                        id INT NOT NULL UNIQUE,
+                        processor CHAR(255) DEFAULT NULL,
+                        PRIMARY KEY (id),
+                        INDEX ready (id, processor)
+                    )
+                    '''), [LOCK_TABLE_NAME]
+                )
+                cursor.execute("INSERT INTO %s (id) SELECT id FROM courseware_studentmodulehistory WHERE id BETWEEN %d AND %d",
+                    [LOCK_TABLE_NAME, self.min_id, self.max_id])
+            except:
+                transaction.rollback()
+            else:
+                transaction.commit()
+
+        logging.info("Migration initialization complete")
```

Comment:
To be clear, an error *will* be thrown and caught. But the result will look the same as a successful initialization.